harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,412 @@
+
+        self._start_ts, self._start_vals = ts, vals
+
+        return {
+            "net_throughput_sent": sent_thru,
+            "net_throughput_recv": recv_thru,
+        }
+
+
+class _Disk(_MetricGroupCollector):
+    group = "disk"
+
+    _disk_paths = ["/", constants.SHARED_FS_CONTAINER_PATH]
+
+    def __init__(self) -> None:
+        # Set initial values for throughput calculations.
+        self._start_ts = time.time()
+        self._start_vals = psutil.disk_io_counters()
+
+        # Initialize accessible disk paths.
+        self._paths = []
+        for path in self._disk_paths:
+            try:
+                psutil.disk_usage(path)
+                self._paths.append(path)
+            except Exception:
+                pass
+
+        super().__init__()
+
+    def collect(self) -> Dict[str, Any]:
+        ts = time.time()
+        vals = psutil.disk_io_counters()
+
+        read_thru = (vals.read_bytes - self._start_vals.read_bytes) / (ts - self._start_ts)
+        write_thru = (vals.write_bytes - self._start_vals.write_bytes) / (ts - self._start_ts)
+        iops = (vals.read_count + vals.write_count) - (
+            self._start_vals.read_count + self._start_vals.write_count
+        ) / (ts - self._start_ts)
+        self._start_ts, self._start_vals = ts, vals
+
+        metrics = {
+            "disk_iops": iops,
+            "disk_throughput_read": read_thru,
+            "disk_throughput_write": write_thru,
+        }
+
+        for path in self._paths:
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> Dict[str, Any]:
+        free_mem_bytes = psutil.virtual_memory().available
+        return {
+            "free_memory": free_mem_bytes / 1e9,
+        }
+
+
+class _CPU(_MetricGroupCollector):
+    group = "cpu"
+
+    def collect(self) -> Dict[str, Any]:
+        cpu_util = psutil.cpu_percent()
+        return {
+            "cpu_util_simple": cpu_util,
+        }
+
+
+class _GPU(_MetricGroupCollector):
+    """
+    Collects GPU metrics for all GPU devices on the machine.
+
+    Returns `gpu_util` and `gpu_free_memory` metrics labeled by the device's UUID. Example:
+
+    .. code::
+        {
+            "GPU-UUID-A": {
+                "gpu_util": 0.12,
+                "gpu_free_memory": 5.662642176,
+            },
+            "GPU-UUID-B": {
+                "gpu_util": 0.23,
+                "gpu_free_memory": 6.552642176,
+            },
+        }
+    """
+
+    group = "gpu"
+
+    def __init__(self):
+        super().__init__()
+
+        self._pynvml_device_handles: Dict[str, Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning(f"pynvml module not found. GPU metrics will not be collected.")
+
+    def _init_pynvml(self) -> None:
+        try:
```

Comment:
An is-not-None assertion here makes my IDE happier.

Reply:
nit: totally trivial but while we're at it, prefer the assertion outside the `try`.